Unknown node type in Marko template: _MarkoEmbed

## Version: 1.2.4

<!--- Provide the exact version in which you see the bug.  You can run `npm ls prettier-plugin-marko` to see this. -->

### Details

Prettifying the following Marko file gives an error

```
<script>
    window.bla = ${JSON.stringify(input.bla)};
</script>
```


### Expected Behavior

File should be formatted

### Actual Behavior

File is not formatted, error is thrown `Unknown node type in Marko template: _MarkoEmbed `

### Your Environment

<!-- Include as many relevant details about the environment you experienced the bug in -->

- MacOS 12.6
- Node v18.13.0
- yarn 1.22.19
- prettier 2.8.4
- @marko/translator-default 5.24.0

### Steps to Reproduce

<!-- Provide a link to a live example -->

<!-- or an unambiguous set of steps to reproduce this bug -->

<!-- include code to reproduce, if relevant -->

1.  Create `index.marko` with the content provided above
2. install prettier and this plugin with yarn
3. install @marko/translator-default (otherwise prettier fails with `Error: [BABEL] index.marko: Cannot find module '@marko/translator-default'`) 
4. run prettier

--- a/src/__tests__/fixtures/embedded-text-with-placeholders.marko
+++ b/src/__tests__/fixtures/embedded-text-with-placeholders.marko
@@ -1,0 +1,8 @@
+<script>
+    window.bla = ${JSON.stringify(input.bla)};
+</script>
+
+<script>
+    window.a = ${JSON.stringify(input.a)};
+    window.b = ${JSON.stringify(input.b)};
+</script>

--- a/src/index.ts
+++ b/src/index.ts
@@ -33,8 +33,9 @@ import {
 } from "./utils/get-original-code";
 
 const defaultFilePath = resolve("index.marko");
-const { builders: b } = doc;
+const { builders: b, utils } = doc;
 const identity = <T>(val: T) => val;
+const embeddedPlaceholderReg = /__EMBEDDED_PLACEHOLDER_(\d+)__/g;
 
 export const languages: SupportLanguage[] = [
   {
@@ -395,59 +396,93 @@ export const printers: Record<string, Printer<Node>> = {
           } else if (node.body.body.length) {
             const lastIndex = node.body.body.length - 1;
             const bodyDocs = [] as Doc[];
-            let textDocs = [] as Doc[];
             let textOnly = true;
 
-            tagPath.each(
-              (child, i) => {
-                const childNode = child.getValue();
-                const isText = isTextLike(childNode, node);
+            if (embedMode) {
+              let placeholderId = 0;
+              const placeholders = [] as Doc[];
+              let embeddedCode = "";
 
-                if (isText) {
-                  textDocs.push(
-                    embedMode && childNode.type === "MarkoText"
-                      ? callEmbed(print, child, embedMode, childNode.value)
-                      : print(child)
-                  );
-                  if (i !== lastIndex) return;
-                } else {
-                  textOnly = false;
-                }
+              tagPath.each(
+                (child) => {
+                  const node = child.getValue();
+                  if (node.type === "MarkoText") {
+                    embeddedCode += node.value;
+                  } else {
+                    embeddedCode += `__EMBEDDED_PLACEHOLDER_${placeholderId++}__`;
+                    placeholders.push(print(child));
+                  }
+                },
+                "body",
+                "body"
+              );
 
-                if (textDocs.length) {
-                  const isFirst = !bodyDocs.length;
-                  bodyDocs.push(
-                    b.group([
-                      opts.markoSyntax === "html"
-                        ? ""
-                        : isFirst
-                        ? b.ifBreak("--", " --", { groupId })
-                        : "--",
-                      opts.markoSyntax === "html"
-                        ? preserveSpace || isFirst
+              const embeddedDoc = replaceEmbeddedPlaceholders(
+                callEmbed(print, tagPath, embedMode, embeddedCode),
+                placeholders
+              );
+
+              bodyDocs.push(
+                b.group([
+                  opts.markoSyntax === "html"
+                    ? ""
+                    : b.ifBreak("--", " --", { groupId }),
+                  opts.markoSyntax === "html" ? "" : b.line,
+                  embeddedDoc,
+                  opts.markoSyntax === "html"
+                    ? ""
+                    : b.ifBreak([b.softline, "--"]),
+                ])
+              );
+            } else {
+              let textDocs = [] as Doc[];
+              tagPath.each(
+                (child, i) => {
+                  const childNode = child.getValue();
+                  const isText = isTextLike(childNode, node);
+
+                  if (isText) {
+                    textDocs.push(print(child));
+                    if (i !== lastIndex) return;
+                  } else {
+                    textOnly = false;
+                  }
+
+                  if (textDocs.length) {
+                    const isFirst = !bodyDocs.length;
+                    bodyDocs.push(
+                      b.group([
+                        opts.markoSyntax === "html"
                           ? ""
-                          : b.softline
-                        : preserveSpace
-                        ? b.hardline
-                        : b.line,
-                      preserveSpace ? textDocs : b.fill(textDocs),
-                      opts.markoSyntax === "html"
-                        ? ""
-                        : b.ifBreak([b.softline, "--"]),
-                    ])
-                  );
+                          : isFirst
+                          ? b.ifBreak("--", " --", { groupId })
+                          : "--",
+                        opts.markoSyntax === "html"
+                          ? preserveSpace || isFirst
+                            ? ""
+                            : b.softline
+                          : preserveSpace
+                          ? b.hardline
+                          : b.line,
+                        preserveSpace ? textDocs : b.fill(textDocs),
+                        opts.markoSyntax === "html"
+                          ? ""
+                          : b.ifBreak([b.softline, "--"]),
+                      ])
+                    );
 
-                  if (!isText) {
-                    textDocs = [];
+                    if (!isText) {
+                      textDocs = [];
+                      bodyDocs.push(print(child));
+                    }
+                  } else {
                     bodyDocs.push(print(child));
                   }
-                } else {
-                  bodyDocs.push(print(child));
-                }
-              },
-              "body",
-              "body"
-            );
+                },
+                "body",
+                "body"
+              );
+            }
 
             const sep =
               (preserveSpace || !textOnly) &&
@@ -708,3 +743,41 @@ export const printers: Record<string, Printer<Node>> = {
     },
   },
 };
+
+function replaceEmbeddedPlaceholders(doc: Doc, placeholders: Doc[]) {
+  if (!placeholders.length) return doc;
+
+  return utils.mapDoc(doc, (cur) => {
+    if (typeof cur === "string") {
+      let match = embeddedPlaceholderReg.exec(cur);
+
+      if (match) {
+        const replacementDocs = [] as Doc[];
+        let index = 0;
+
+        do {
+          const placeholderIndex = +match[1];
+
+          if (index !== match.index) {
+            replacementDocs.push(cur.slice(index, match.index));
+          }
+
+          replacementDocs.push(placeholders[placeholderIndex]);
+          index = match.index + match[0].length;
+        } while ((match = embeddedPlaceholderReg.exec(cur)));
+
+        if (index !== cur.length) {
+          replacementDocs.push(cur.slice(index));
+        }
+
+        if (replacementDocs.length === 1) {
+          return replacementDocs[0];
+        }
+
+        return replacementDocs;
+      }
+    }
+
+    return cur;
+  });
+}

